### ISSUE TITLE
checkers rules

### ISSUE THREAD

**Issue by gilberto85** [stillonearth/CheckersOnBevy#3]
i dont know but i saw your video and this is not american checkers rules 

https://www.itsyourturn.com/t_helptopic2030.html

**Comment by stillonearth** [comment]
Those should be English Checkers. Do you need American checkers?

**Comment by gilberto85** [comment]
yes

**Comment by stillonearth** [comment]
Check the updated video. Rules as of now:

- getting an extra turn for a piece if you ate an opponent
- crowning

**Comment by gilberto85** [comment]
Hi there thank you I saw the video but american checkers jumps are mandatory

**Comment by stillonearth** [comment]
They are. Check 16th second on vid

**Comment by gilberto85** [comment]
Sorry the only video I see is this 
Checkers.on.Bevy.2022-11-16.19-08-04.mp4


And its not jumping according to American checkers rules btw you have a very good skills

**Comment by stillonearth** [comment]
Yeah, that's the one

Where exactly you see an error in this ruleset:

- piece can move 1 square in advancing forward
- piece get an extra move if it eaten an opponent's piece
- after crowning a piece can move diagonally without restriction on number of squares

**Comment by gilberto85** [comment]
If a piece is in front of another piece with a free square nehide him it
need to jump and remove the piece from the board ( jumps are mandatory ) if
you want you can see a video here

Look everytime they need to jump in a piece that piece needs to be removed
from the board

https://youtube.com/shorts/-TnVXMNsoHI?feature=share

**Comment by stillonearth** [comment]
Okay, seems I'm getting it now.

**Comment by gilberto85** [comment]
Hope that video can help you

**Comment by gilberto85** [comment]
Also multiple jumps are mandatory

**Comment by gilberto85** [comment]
Is that much to change ?  I don't know what kind of checkers is that rules now I don't know if it's Italian rules

**Comment by stillonearth** [comment]
No, it's not that much, I'll aim to update them this week.

пт, 18 нояб. 2022 г., 04:21 gilberto85 ***@***.***>:

**Comment by gilberto85** [comment]
I am working on my own checkers website , the developer make something
already and we can play already and creat rooms to play against other
people. There are some bugs that I need to fix and some other checkers
variants that I need to add , would you be interested in working on my site
?

Checkersvip.com

**Comment by stillonearth** [comment]
This project supports wasm targets so you can depoy on web (though neural AI won't work with wasm, an investigation required). With some adjustments multiplayer is also possible. I'll make an example branch with american rules adjustments.

PRs are welcome if you'll choose to use this codebase.

**Comment by stillonearth** [comment]
This branch implements American ruleset (i.e. jump mandatory is possible): https://github.com/stillonearth/CheckersOnBevy/tree/american-ruleset

The changes are in ```pub fn possible_moves(&self) -> [Vec<Position>; 24]```:

1. Check whether current moveset for a given side contains a 'Take` move:
```rust
let mut american_jump_possible: HashMap<Color, bool> = HashMap::new();
for i in 0..24 {
    let p = self.state.pieces.iter().find(|p| p.id == i);
    if p.is_none() {
        continue;
    }

    let color = p.unwrap().color;
    for j in 0..24 {
        let p = self.state.pieces.iter().find(|p| p.id == j);
        if p.is_none() {
            continue;
        }

        let other_color = p.unwrap().color;

        if other_color == color {
            let jump_possible = moveset[j as usize].iter().any(|m| m.1 == MoveType::Take);
            if jump_possible {
                american_jump_possible.insert(color, true);
                break;
            }
        }
    }
}
```

2. Filter out regular moves if so.

That's enough to limit AI, and there's another check in ```Game::step```

```rust
let moveset = self.possible_moves();
if moveset[piece.id as usize].len() == 0 {
    // self.state.turn.change();
    return (MoveType::Invalid, &self.state, self.check_termination());
}
```

**Comment by gilberto85** [comment]
I saw the last video and you improved the code however keep in mind that of
you or the oponent have multiple jumps those are mandatory as well

**Comment by gilberto85** [comment]
On second 11' of that video the AI had a chance make another jump and it
waited until you moved to make the other jump

**Comment by gilberto85** [comment]
My website supports playonline already but I need users to play against an
AI as well or creat bots so if a user creat a room a no one join the room
in let's say 30 seconds a bot will join and play with the real user.

**Comment by stillonearth** [comment]
You're talking about this vid right? https://github.com/stillonearth/CheckersOnBevy/tree/american-ruleset

In that branches attack moves are mandatory.

**Comment by gilberto85** [comment]
Yes that video if you have multiple jumps ( 2, 3 , 4 etc ) are mandatory on
the same turn they not need to waiti until the oponen moves

**Comment by stillonearth** [comment]
That's probably an animation issue, as they are not queued now.
